dev: add demo server & pack to docker

asd-a opened the pull request:
#3

asd-a committed:
dev: demo server

asd-a committed:
dev: pack to docker

asd-a committed:
fix: code quality

asd-a committed:
fix: docker security

CoderOJ commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+FROM ubuntu:22.04 as build
```
use a rust container instead of install rust

CoderOJ commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+# Prepare the environment
+
+RUN racket api/rust.rkt < api/api.rkt > api/src/lib.rs
```
use makefile inside api repo, see se-clavier/api#5

CoderOJ commented on src/main.rs:
```diff
@@ -1,4 +1,44 @@
+            message: "token".to_string(),
+        })
+    }
+    async fn login2(&mut self, _req: api::LoginRequest) -> Result<api::LoginResponse2, api::Error> {
```
update api with se-clavier/api#4

CoderOJ commented:
And check sonarqube issue (use `AS` instead of `as` in Dockerfile).

CoderOJ commented on Cargo.toml:
```diff
@@ -4,3 +4,9 @@ version = "0.1.0"
+tokio = { version = "1.43.0", features = ["full"] }
+axum = { version = "0.8.1", features = ["macros"] }
+serde_json = "1.0.140"
+typetag = "0.2.20"
```
remove this if unnecessary

CoderOJ commented on src/main.rs:
```diff
@@ -1,4 +1,44 @@
+async fn handler(
+    State(mut app): State<App>,
+    Json(body): Json<APICollection>,
+) -> Result<Json<impl Serialize>, Json<api::Error>> {
```
Is the `status_code` set properly here by returning `Result<Json, Json>`? I think a better approach is to implement `IntoResponse` for `Result<Box<dyn Serialize>, api:ERROR>`

CoderOJ commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+# Clone the repository
+
+RUN git clone https://github.com/se-clavier/back-end.git .
```
cloning repo / submodule update should not be included in Dockerfile

CoderOJ replied:
Instead, use `COPY` to copy files in current directory and use `.dockerignore` to ignore files.

asd-a committed:
dev: build api with makefile & change build image to rust image

asd-a committed:
fix: remove unnecessary deps

asd-a committed:
feat(docker): add image version

CoderOJ commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM rust:1.85 AS build
```
use rust:1.85-bookworm

asd-a committed:
feat(status_code): add status code to error response

asd-a committed:
feat(docker): use curruent commit to build image

asd-a committed:
fix(docker): build issues & security hotspots

- remove recursively copying context directories
- fix debian image tag
- fix user missing issue
- fix path traversal issue

asd-a committed:
fix(docker): sepecify path pattern

asd-a committed:
fix(docker): remove glob pattern

asd-a committed:
dep(api): bump api version
